packages/pds/src/db/message-queue/index.ts

```diff
@@ -0,0 +1,276 @@
+      .returning(['did', 'count'])
+      .execute()
+  }
+
+  private async handleAddUpvote(
+    db: Database,
+    message: AddUpvote,
+  ): Promise<void> {
+    const userScenes = await db.getScenesForUser(message.user)
+
+    const res = await db.db
+      .updateTable('scene_votes_on_post')
+      .set({ count: sql`count + 1` })
+      .where('subject', '=', message.subject)
+      .where('did', 'in', userScenes)
+      .returning(['did', 'count', 'postedTrending'])
+      .execute()
+
+    const toInsert: SceneVotesOnPost[] = userScenes
+      .filter((scene) => !res.some((row) => scene === row.did))
+      .map((scene) => ({
+        did: scene,
+        subject: message.subject,
+        count: 1,
+        postedTrending: 0,
+      }))
+
+    if (toInsert.length > 0) {
+      await db.db.insertInto('scene_votes_on_post').values(toInsert).execute()
+    }
+    await this.checkTrending(db, userScenes, message.subject)
+  }
+
+  private async handleRemoveUpvote(
+    db: Database,
+    message: RemoveUpvote,
+  ): Promise<void> {
+    const userScenes = await db.getScenesForUser(message.user)
+    await db.db
+      .updateTable('scene_votes_on_post')
+      .set({ count: sql`count + 1` })
+      .where('subject', '=', message.subject)
+      .where('did', 'in', userScenes)
+      .returning(['did', 'count'])
+      .execute()
+  }
+
+  private async handleCreateNotification(
+    db: Database,
+    message: CreateNotification,
+  ): Promise<void> {
+    await db.db.insertInto('user_notification').values({
+      userDid: message.userDid,
+      recordUri: message.recordUri,
+      recordCid: message.recordCid,
+      author: message.author,
+      reason: message.reason,
+      reasonSubject: message.reasonSubject,
+      indexedAt: new Date().toISOString(),
+    })
+  }
+
+  private async handleDeleteNotifications(
+    db: Database,
+    message: DeleteNotifications,
+  ) {
+    await db.db
+      .deleteFrom('user_notification')
+      .where('recordUri', '=', message.recordUri)
+      .execute()
+  }
+
+  // Side effects
+  // -------------
+  private async checkTrending(
+    db: Database,
+    scenes: string[],
+    subject: string,
+  ): Promise<void> {
+    const state = await db.db
+      .selectFrom('scene_member_count as members')
+      .innerJoin('scene_votes_on_post as votes', 'votes.did', 'members.did')
+      .innerJoin('post', 'post.uri', 'votes.subject')
+      .where('members.did', 'in', scenes)
+      .where('votes.subject', '=', subject)
+      .select([
+        'members.did as did',
+        'members.count as memberCount',
+        'votes.subject as subject',
+        'post.cid as subjectCid',
+        'votes.count as voteCount',
+        'votes.postedTrending as postedTrending',
+      ])
+      .execute()
+
+    const now = new Date().toISOString()
+    await Promise.all(
+      state.map(async (scene) => {
+        if (scene.postedTrending) return
+        const ratio = scene.voteCount / scene.memberCount
```

Comment:
Probably want to sanity check against a divide by zero here. Something is very wrong if that happens but you know

Reply:
good catch 👌